oama: init at 0.14

aidalgol opened the pull request:
## Description of changes
Adds `oama` and remove `mailctl`, because `mailctl` upstream has been replaced by `oama`.
https://github.com/pdobsan/oama/discussions/26

## Things done
- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

aidalgol committed:
mailctl: remove

aidalgol commented:
The code-formatting check failed on the new file in this PR, but when I run `nixpkgs-fmt` locally on the file, it says it is already formatted.

cdepillabout commented on pkgs/tools/networking/oama/default.nix:
A couple things here:

1. Is there anyway to get upstream to release this to Hackage?  Updating Haskell packages in Nixpkgs (and other Linux distros) happens all automatically for packages that are released to Hackage, but not for packages that are just in GitHub.
2. If the upstream author is not interested in releasing to Hackage, I find it generally nicer to split the `cabal2nix` output from the hand-edited portion of the derivation, and include an update script.  Take a look at `pkgs/development/tools/haskell/dconf2nix/` or `pkgs/development/tools/purescript/spago/` for examples of this.  This should also help you get rid of the manual overrides in `pkgs/top-level/all-packages.nix`.

cdepillabout commented:
> Adds oama and remove mailctl, because mailctl upstream has been replaced by oama.

I don't have any context here, but is there something you can link to upstream that officially says `oama` has replaced `mailctl`?

> The code-formatting check failed on the new file in this PR, but when I run nixpkgs-fmt locally on the file, it says it is already formatted.

I'm vaguely aware that there are two different formatters, both called something like `nixpkgs-fmt`.  Are you sure you're using the right one?

aidalgol commented:
> > Adds oama and remove mailctl, because mailctl upstream has been replaced by oama.
> 
> I don't have any context here, but is there something you can link to upstream that officially says `oama` has replaced `mailctl`?

No, just that the `mailctl` repo redirects to the `oama` repo.

> > The code-formatting check failed on the new file in this PR, but when I run nixpkgs-fmt locally on the file, it says it is already formatted.
> 
> I'm vaguely aware that there are two different formatters, both called something like `nixpkgs-fmt`. Are you sure you're using the right one?

I'm using the executable named `nixpkgs-fmt` from the derivation `nixpkgs-fmt`.

s0me1newithhand7s commented on pkgs/tools/networking/oama/default.nix:
```diff
@@ -0,0 +1,97 @@
+    warp
+    yaml
+  ];
+
```
isn't here needed to be:
```Nix
meta = with lib; {
<> 
};
```
?

aidalgol replied:
The `haskellPackages` version of `mkDerivation` doesn't like that, and instead all the attributes that would normally go in `meta` go at the top level.

cdepillabout commented:
@aidalgol 

> No, just that the mailctl repo redirects to the oama repo.

Could you link to that in this PR, just for posterity?

> I'm using the executable named nixpkgs-fmt from the derivation nixpkgs-fmt.

Please look at the error message in the GitHub action, it tells you what to do.

maralorn commented:
Code formatting needs to happen with `nixfmt` not `nixpkgsfmt`.

maralorn commented:
Secondly, if this is not on hackage we should put this in the pkgs/by-name hierarchy.

aidalgol committed:
oama: init at 0.14

aidalgol commented:
How does this look now?  I would like to add an update script before this is merged, but first wanted to get feedback on the package rewrite.

I also managed to find a note about `mailctl` becoming `oama` in the git history, and added a link to the referenced GitHub Discussions thread with the background, which I have put in the PR description.  The redirect is because it's the same repo, just renamed.  That also means that this tool used to be published on Hackage (it's gone now, but I distinctly recall looking at the dependencies of `mailctl` from its page on Hackage), so it appears the author has moved to publishing via GitHub releases instead.

maralorn commented on pkgs/top-level/aliases.nix:
```diff
@@ -921,6 +921,7 @@ mapAliases ({
 
   ma1sd = throw "ma1sd was dropped as it is unmaintained"; # Added 2024-07-10
   MACS2 = macs2; # Added 2023-06-12
+  mailctl = throw "mailctl has been supseseded upstream by oama"; # Added 2024-08-19
```
```suggestion
  mailctl = throw "mailctl has been superseeded upstream by oama"; # Added 2024-08-19
```

maralorn reviewed (commented):
LGTM

croissong commented:
FYI there's already an unmerged PR for the same package: https://github.com/NixOS/nixpkgs/pull/319267

maralorn commented:
Oh, no. How sad. The Haskell team wasn’t notified there, so I didn’t know.

aidalgol commented:
Added an update script and used it to regenerate the `generated-package.nix` file.  The formatting PR check is complaining about the generated file because I am not running the output of `cabal2nix` through the formatter in the update script, based on @maralorn's earlier comment.
> However for cabal2nix generated files I think we should figure out how to opt-out of filtering.

cdepillabout commented on pkgs/by-name/oa/oama/update.sh:
```diff
@@ -0,0 +1,21 @@
+derivation_file="${script_dir}/generated-package.nix"
+latest_version="$(curl --silent https://api.github.com/repos/pdobsan/oama/releases/latest | jq --raw-output '.tag_name')"
+
+echo "Updating nixfmt-rfc-style to version ${latest_version}."
```
```suggestion
echo "Updating oama to version ${latest_version}."
```

aidalgol replied:
Oops!

cdepillabout commented:
> The formatting PR check is complaining about the generated file because I am not running the output of cabal2nix through the formatter in the update script

I imagine it would be easier to just run the output of `cabal2nix` through `nixfmt`, unless someone can suggest a way to opt-out of formatting.

Other than that, this LGTM.

aidalgol commented:
The ofborg-eval failure looks unrelated to these changes.

cdepillabout commented:
@ofborg build oama

cdepillabout commented:
Thanks, this LGTM!

cdepillabout commented:
@aidalgol Also, I'd encourage you to send a follow-up PR to the `haskell-updates` branch that adds `oama` to this list:

https://github.com/NixOS/nixpkgs/blob/0e4fd9c60f49174f745cba3659a89c3b202e0f84/pkgs/top-level/release-haskell.nix#L266-L279

That way you'll get pinged on GitHub if `oama` ever stops building on our `haskell-updates` branch.
